File: polyformalism-languages/conservation-law/conservation.zig
// ZIG: What memory layout reveals about conservation
// Forces you to think about: Integer types, exact width, explicit conversion.
//   u32 for counting, f64 for computing — these are different mathematical objects.
// Broken assumption: That domain violations (n=0) are runtime problems.
//   Zig forces confronting the mathematical domain at the type level.
// Novel insight: comptime vs runtime — the domain of δ(n) can be checked at compile time.

const std = @import("std");
const math = std.math;

fn delta(n: f64) f64 {
    return (1.0 / math.sqrt(n)) * (1.0 - 3.0 / (2.0 * n));
}

pub fn main() void {
    std.debug.print("    n       δ(n)     CCR%\n", .{});
    std.debug.print("{s}\n", .{"-" ** 28});
    var i: u32 = 1;
    while (i <= 100) : (i += 1) {
        const d = delta(@floatFromInt(i));
        std.debug.print("{d:>5}  {d:>10.6}  {d:>7.2}%\n", .{ i, d, d * 100.0 });
    }
}

File: polyformalism-languages/conservation-law/conservation.jl
# JULIA: What multiple dispatch reveals about conservation
# Forces you to think about: Mathematical notation IS syntax. Why should δ(n) look
#   different in code than in a paper? It shouldn't.
# Broken assumption: That scalar and vector operations are categorically different.
# Novel insight: δ.(1:100) — the dot broadcast makes "apply to all fleet sizes"
#   syntactically invisible. The fleet is an array primitive.

δ(n) = (1 / √n) * (1 - 3 / (2n))
ccr(n) = δ(n) * 100

println("    n       δ(n)     CCR%")
println("-" ^ 28)
for n in 1:100
    println(lpad(n, 5), "  ", lpad(round(δ(n), digits=6), 10), "  ", lpad(round(ccr(n), digits=2), 7), "%")
end
